File: Google/1438 Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetCode
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var solution = new Solution();
            solution.LongestSubarray(new int[] { 9, 8, 4, 2 }, 5);
        }
    }

    public class Solution
    {
        public int LongestSubarray(int[] nums, int limit)
        {
            if (nums == null || nums.Length < 1 || limit < 0)
            {
                throw new ArgumentException();
            }

            var max = new LinkedList<int>();
            var min = new LinkedList<int>();
            var left = 0;
            var right = 0;
            
            while (right < nums.Length)
            {
                var num = nums[right++];
                while (max.Count > 0 && num > max.Last.Value) max.RemoveLast();
                while (min.Count > 0 && num < min.Last.Value) min.RemoveLast();

                max.AddLast(num);
                min.AddLast(num);

                if (max.First.Value - min.First.Value <= limit) continue;

                num = nums[left++];
                if (max.First.Value == num) max.RemoveFirst();
                if (min.First.Value == num) min.RemoveFirst();
            }

            return right - left;
        }

        public int LongestSubarraySlow(int[] nums, int limit)
        {
            if (nums == null || nums.Length < 1 || limit < 0)
            {
                throw new ArgumentException();
            }

            var minPos = 0;
            var maxPos = 0;

            var start = 0;
            var longest = 1;
            var length = 1;
            for (var i = 1; i < nums.Length; i++)
            {
                var value = nums[i];
                var diffMin = Math.Abs(value - nums[minPos]);
                var diffMax = Math.Abs(value - nums[maxPos]);

                if (diffMin > limit || diffMax > limit)
                {
                    longest = Math.Max(i - start, longest);

                    var pos = -1;
                    if (diffMin > limit)
                        pos = minPos;

                    if (diffMax > limit)
                        pos = Math.Max(pos, maxPos);

                    start = i;
                    minPos = i;
                    maxPos = i;

                    var curr = i;
                    while (--curr > pos)
                    {
                        var diff = Math.Abs(nums[curr] - value);
                        if (diff > limit)
                            break;

                        start = curr;
                        if (nums[curr] < nums[minPos])
                            minPos = curr;
                        else if (nums[curr] > nums[maxPos])
                            maxPos = curr;
                    }

                    length = i - start + 1;
                }
                else
                {
                    length++;

                    if (value < nums[minPos])
                        minPos = i;
                    else if (value > nums[maxPos])
                        maxPos = i;
                }
            }

            return Math.Max(longest, length);
        }
    }
}

File: Google/1438 Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit.js
/**
 * @param {number[]} nums
 * @param {number} limit
 * @return {number}
 */
var longestSubarray = function(nums, limit) {
    let max = new LinkedList();
    let min = new LinkedList();
    let left = 0;
    let right = 0;

    while (right < nums.length) {
    	let num = nums[right++];
    	while (max.count > 0 && max.last.value < num) max.removeLast();
    	while (min.count > 0 && min.last.value > num) min.removeLast();

    	max.addLast(num);
    	min.addLast(num);

    	if (max.first.value - min.first.value <= limit) continue;

    	num = nums[left++];
    	if (max.first.value === num) max.removeFirst();
    	if (min.first.value === num) min.removeFirst();
    }

    return right - left;
};

class LinkedList {
	constructor() {
		this.head = null;
		this.tail = null;
		this.count = 0;
	}

	get first() {
		return this.head;
	}

	get last() {
		return this.tail;
	}

	addLast(num) {
		let node = new Node(num);

		if (!!this.tail) {
			this.tail.next = node;
			node.previous = this.tail;
		}

		this.count++;
		this.tail = node;
		if (!this.head) {
			this.head = node;
		}
	}

	removeLast() {
		if (this.count < 1) {
			return;
		}

		if (this.count === 1) {
			this.head = this.tail = null;
		} else {
			this.tail = this.tail.previous;
			this.tail.next = null;
		}
		this.count--;
	}

	removeFirst() {
		if (this.count < 1) {
			return;
		}

		if (this.count === 1) {
			this.head = this.tail = null;
		} else {
			let node = this.head.next;
			node.previous = null;
			this.head.next = null;
			this.head = node;
		}
		this.count--;
	}
}

class Node {
	constructor(value, next = null, previous = null) {
		this.value = value;
		this.next = next;
		this.previous = previous;
	}
}

longestSubarray([8, 2, 4, 7], 4);